plugins/modules/network_compliance_workflow_manager.py

```diff
@@ -1304,37 +1238,162 @@ def get_compliance_task_status(self, task_id, mgmt_ip_instance_id_map):
-                self.update_result("failed", False, self.msg, "CRITICAL")
-                break
+            # Append the current batch result to the batches_result list
+            batches_result.append(batch_result)
+
+        self.log("Collective result of all batches: {0}".format(batches_result), "DEBUG")
+        return batches_result
+
+    def validate_batch_result(self, batches_result, retried_batches=None):
+        """
+        Validates the results of compliance check tasks for device batches.
+        Parameters:
+            batches_status (list): A list of dictionaries where each dictionary contains 'task_id',
+                                   'batch_params', 'task_status', and 'msg' for each batch.
+        Returns:
+            list: A list of device IDs that have successfully completed the compliance check.
+        Description:
+            This function iterates over the provided batches status, checks the task status for each batch,
+            and collects the device IDs for batches that have completed successfully. For batches that failed,
+            it re-runs the compliance check with a batch size of 1, validates the results recursively, and collects
+            the successful device IDs.
+        """
+        if retried_batches is None:
+            retried_batches = set()
+        successful_devices = []
+
+        # Iterate over each batch in the batches_status list
+        for batch in batches_result:
+            task_status = batch.get("task_status")
+            batch_params = batch.get("batch_params")
+            device_ids = tuple(batch_params.get("deviceUuids"))
+
+            # Check if the task status is successful
+            if task_status == "Success":
+                successful_devices.extend(device_ids)
+            else:
+                # Check if the batch has already been retried with batch size of 1
+                if device_ids in retried_batches:
+                    self.log(
+                        "Batch for device(s) {0} has already been retried with batch size of 1 and failed. "
+                        "Stopping recursion.".format(", ".join(device_ids)),
+                        "ERROR"
+                    )
+                    continue
+
+                self.log("Re-running compliance check for batch {0} with batch_result: {1} ".format(batch, batches_result), "WARNING")
+                # Re-run the compliance check for the failed batch with batch size of 1
+                retried_batches.add(device_ids)
+                batches_dict = self.run_compliance(batch_params, batch_size=1)
+                batches_result = self.get_batches_result(batches_dict)
+
+                # Recursively validate the batch results and append the successful device IDs
+                successful_devices.extend(self.validate_batch_result(batches_result, retried_batches=retried_batches))
+        return successful_devices
```

Comment:
Can we add a new line before return? also pint a log message of successful_devices before returning?